tests/ast/ast_test_utils.py

```diff
@@ -0,0 +1,253 @@
+    error_msg = "\nOriginal stack trace:\n" + "".join(
+        filter(
+            lambda s: "site-package" not in s and "ast_test_utils" not in s,
+            traceback.format_stack(),
+        )
+    )
+
+    # Include the generated code in the result.
+    error_msg = "*" * 80 + "\n" + python_text + "\n" + "*" * 80 + "\n" + error_msg
+
+    return error_msg
+
+
+def notify_compare_ast_validation_with_session(
+    validation_session: Session, query_record: QueryRecord, *args, **kwargs
+):
+    # Get query_id for the original and generated code python dataframe code.
+    qid_result1 = validation_session._ast_full_validation_result
+    qid_result2 = query_record.query_id
+
+    # Compare the results from both queries, if 0 then equal otherwise not equal.
+    comparison_sql = f"""
+        select sum(*) from (
+            select sum(*) as result1 from (select hash(*) from table(result_scan('{qid_result1}')))
+            union all
+            select -sum(*) as result2 from (select hash(*) from table(result_scan('{qid_result2}')))
+        )
+    """
+
+    error_msg = ""
+    success = False
+    try:
+        results_cursor = validation_session._conn._cursor.execute(comparison_sql)
+        compare_results = validation_session._conn._to_data_or_iter(
+            results_cursor, False, False
+        )
+        compare_diff = compare_results["data"][0][0]
+        success = compare_diff == 0
+        if not success:
+            error_msg = f"Full AST validation results differed.\n\nResult 1 Query Id: {qid_result1}\nResult 2 Query Id: {qid_result2}\n"
+    except Exception as ex:
+        error_msg = str(ex) + "\n"
+        success = False
+
+    if not success:
+        error_msg = error_msg + generate_error_trace_info(
+            validation_session._debug_python_code_output
+        )
+
+    pytest.assume(
+        success,
+        f"""Full AST validation failed.\n{error_msg}""",
+    )
+
+
+def notify_full_ast_validation_with_listener(
+    full_ast_validation_listener: AstListener,
+    query_record: QueryRecord,
+    *args,
+    **kwargs,
+):
+    # For tests that contain multiple actions, it's possible a later action depends on some dataframe code that
+    # was provided as part of an earlier action.  Therefore we must keep all the previous statements and replay
+    # them (excluding prior actions that already executed) in case there is a dependency.  This ensures the unparser
+    # can reconstruct the full python code for execution.
+
+    # Call the original listener notify method
+    full_ast_validation_listener._original_notify(query_record, *args, **kwargs)
+
+    # Obtain the test name from the scope of the test method.
+    test_name = os.environ.get("PYTEST_CURRENT_TEST")
+
+    # If this is part of the setup then we can skip, as we don't want to run twice in validation mode.  Some setup
+    # code directly calls internal functions that circumvent the AST path.
+    if "(setup)" in test_name:
```

Comment:
should we return if `test_name is None` as well?

Reply:
It shouldn't be None as pytest adds this env variable, but can add a check anyway just in case.